add reload() for QMeshLayer

vcloarec opened the pull request:
## Description
Implementation of :
- QMeshLayer::reload()
- QgsMdalProvider::reloadData()

- Add test in qmeshlayertest.cpp.

This PR may open the possibility to modify the mesh by external application or plugin when it is open in QGis.

Commit:
add reload() for QMeshLayer

PeterPetrik commented on tests/src/core/testqgsmeshlayer.cpp:
```diff
@@ -493,5 +495,71 @@ void TestQgsMeshLayer::test_time_format()
+  QCOMPARE( 8, layer.nativeMesh()->vertexCount() );
+  QCOMPARE( 5, layer.nativeMesh()->faceCount() );
+
+
```
remove extra empty lines in all files

PeterPetrik commented on tests/src/core/testqgsmeshlayer.cpp:
```diff
@@ -493,5 +495,71 @@ void TestQgsMeshLayer::test_time_format()
   QCOMPARE( time, expectedTime );
 }
 
+void TestQgsMeshLayer::test_reload()
```
can you add test with some datasets assigned?

PeterPetrik commented on src/providers/mdal/qgsmdalprovider.cpp:
```diff
@@ -241,6 +241,22 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
 
+void QgsMdalProvider::reloadData()
+{
+  if ( mMeshH )
```
I think you need to reset mNativeMesh, mTriangularMesh and mRendererCache

vcloarec replied:
mNativeMesh and mRendererCache is reset in the new reload() method.
mTriangularMesh will be updated when createRendrer() is called, no ?

PeterPetrik replied:
what is "new reload" method? I would just reset with NULL for all three

PeterPetrik commented on src/providers/mdal/qgsmdalprovider.cpp:
```diff
@@ -241,6 +241,22 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
+    MDAL_CloseMesh( mMeshH );
+
+  QByteArray curi = dataSourceUri().toUtf8();
+  mMeshH = MDAL_LoadMesh( curi.constData() );
```
maybe we can reuse setDataProvider() instead of this code?

vcloarec replied:
?

PeterPetrik replied:
just that this logic is already in the `setDataProvider()` method, so I am wondering if we cannot reuse it (maybe not)

vcloarec replied:
To avoid misunderstanding : do you refer to the code in QgsMeshLayer::reload() or to the code is commented ? (I have to ask you the same question for the previous comment).

For the QgsMdalProvider class, I think we can now factor the reloadData method and the constructor with a loadMesh method to avoid duplicate code.

vcloarec replied:
> maybe we can reuse setDataProvider() instead of this code?

setDataProvider() destruct the current provider and create a new one. It seems to be a quick way to reset and reload data.
But I think we should give the responsibility to the provider for choosing the way the data could be reloaded.
We don't know how futur providers will work to access data. It could be a problem if the provider instance is destructed and replaced by a new one, especially if the provider is link with a extern mechanism. What do you think about ?

PeterPetrik replied:
agree

PeterPetrik commented on src/providers/mdal/qgsmdalprovider.cpp:
```diff
@@ -241,6 +241,22 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
+    const QString proj = MDAL_M_projection( mMeshH );
+    if ( !proj.isEmpty() )
+      mCrs.createFromString( proj );
+  }
```
what happens if mesh layer has assigned some external datasets? what if they are no longer compatible with the mesh?

vcloarec replied:
may be a control to check vertices count and faces coun,. If not the same, external datasets are removed ?

PeterPetrik replied:
that should do the job

vcloarec replied:
After consideration about extra datasets, and sorry for this long post :

It seems the control of the extra dataset files is the responsability of the provider.

For the memory provider, maybe the reload function doesn't have sense, so, I purpose to not re-implement the reloadData() method.

For the MDAL provider, it is the MDAL API that loads and control the dataset. So when the mesh is not compatible, the behavior depends of each MDAL's drivers ... I don't have analyze the code of all the drivers, but it seems if the mesh is not compatible, the dataset is not added in the MDAL::Mesh. Can you confirm?

So, when mesh is reloaded, if a extra dataset file is no valid or the mesh is not compatible, to synchronize the Dataset stored in the MDAL::Mesh, the simple choice is to remove the uri of the file from mExtraDatasetUris list.

But I am not sure it is the best choice for the user because the extra dataset will disappear from QGIS and can't come back if the file return to be valid or compatible, except if the user reload it.

Maybe, it is more convenient for the user if the uri of the file is kept in the mExtraDatasetUris list, so its reload would be tried again during the next reload of the layer. But it is necessary to synchronize this list with the list of DatasetGroup stored in the MDAL::Mesh. To do that, I would purpose to add a void DatasetGroup in the MDAL::Mesh when QgsMdalProvider::reloadData() method is called and the extra Dataset file is not valid or incompatible. But I don't find a way to do that without modify the MDAL API... Maybe I am wrong?

As I can't implement the second solution directly in QGIS code, I retained the first one. I have added tests for this solution (with an new test data file).

PeterPetrik commented:
Hi, this approach looks good!

vcloarec committed:
in QgsMeshLayer:eload() : reset mTriangularMesh

QgsMdalProvider :
  refactor constructor and reloadData() in a loadData() method
  add extra dataset loading

Add test for qgsmeshlayer

vcloarec committed:
add test data file

PeterPetrik commented on src/providers/mdal/qgsmdalprovider.cpp:
```diff
@@ -241,6 +234,39 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
+        mExtraDatasetUris.removeOne( uri );
+      }
+    }
+
```
I think you need to emit `QgsMeshDataProvider::dataChanged` signal here to notify GUI to reload widgets and refresh canvas. 

What I am not sure is about removal of extra datasets. So far we have just signal `datasetGroupsAdded`, but maybe we need also `datasetGroupsRemoved` to emit here to notify that we removed X datasets so the dataset group tree can be properly refreshed too.

Can you please try

1. load mesh with extra dataset (so bed elevation + extra one)
2. change mesh so extra dataset is not valid anymore
3. check the mesh layer properties to see if you have only bed elevation dataset loaded

vcloarec replied:
> I think you need to emit `QgsMeshDataProvider::dataChanged` signal here to notify GUI to reload widgets and refresh canvas.

There is a problem :
QgsDataProvider::dataChanged is connect with refreshMapCanvas
```suggestion
https://github.com/qgis/QGIS/blob/4aaa523977c9ec74ddb2e92f704eaa30d16d5121/src/app/qgisapp.cpp#L12346-L12350
```
This method call QgsMapCanvas::refreshAllLAyer (line 9401 same file)
```suggestion
https://github.com/qgis/QGIS/blob/4aaa523977c9ec74ddb2e92f704eaa30d16d5121/src/app/qgisapp.cpp#L9394-L9401
```
QgsMapCanvas::refreshAllLayer call QgsMapLayer::reload :
```suggestion
https://github.com/qgis/QGIS/blob/4aaa523977c9ec74ddb2e92f704eaa30d16d5121/src/gui/qgsmapcanvas.cpp#L2188-L2194
```
and QgsMeshLayer::reload() call QgsMeshDataProvider::reloadData() making a **circular reference**

Is it really necessary to emit QgsMeshDataProvider::dataChanged from here ? 
May the caller of QgsMeshLayer::reload() or QgsMeshDataProvider::reloadData() can keep the responsability to emit dataChanged ?

vcloarec replied:
> Can you please try
> 
>     1. load mesh with extra dataset (so bed elevation + extra one)
> 
>     2. change mesh so extra dataset is not valid anymore
> 
>     3. check the mesh layer properties to see if you have only bed elevation dataset loaded

Something similar has been done in the added test, no ?

vcloarec replied:
> I think you need to emit `QgsMeshDataProvider::dataChanged` signal here to notify GUI to reload widgets and refresh canvas.

There is a problem :
`QgsDataProvider::dataChanged` is connect with `refreshMapCanvas` (line 12349 in qgisapp.cpp)

This method calls `QgsMapCanvas::refreshAllLayers` (line 9401 same file)

`QgsMapCanvas::refreshAllLayer()` calls `QgsMapLayer::reload() `

and `QgsMeshLayer::reload()` calls `QgsMeshDataProvider::reloadData()` 

making a **circular reference**

Is it really necessary to emit `QgsMeshDataProvider::dataChanged` from here? 
May the caller of `QgsMeshLayer::reload()` or `QgsMeshDataProvider::reloadData()` can keep the responsability to inform the data has changed when necessary ?

vcloarec replied:
> What I am not sure is about removal of extra datasets. So far we have just signal `datasetGroupsAdded`, but maybe we need also `datasetGroupsRemoved` to emit here to notify that we removed X datasets so the dataset group tree can be properly refreshed too.

Remove extra dataSets is not as simple it looks like... I look at the code of the class  `QgsMeshRendererSettings` and the renderer settings are stored in a QHash with index of the groupsDataset as key. If we remove extra dataset, the index will change... Maybe  we need to store the renderer settings in QList to make easier the synchronization with the `mExtraDatasetUris` list.
It make me thinking about the solution I have purpose sooner: not remove the datasetGroup if it is no valide or incompatible but replace it by a void datasetGroup. So no need to delete the datasetGroup...
On the other hand, It is interesting to open the possibility to remove extra datasets : the user may wants to remove extra datasets he doesn't want its anymore.
What do you think about ?

PeterPetrik replied:
ok with `dataChanged()`. we just need to make sure that this `reload()` is not called with every refresh of the canvas, which is not the case (Recreating the mesh/triangular mesh can take ages for large datasets)

as for `mExtraDatasetUris` it is even more complicated because one extra URI can have multiple dataset groups in it (e.g. XMDF format or so).

it would be good to have possibility to remove datasets from gui, but that would need some work in MDAL too (to have some kind of UUID of datasets or so) so we can persist in QGIS project which datasets should not be loaded. 

for now I would just simply add there another function similar to `QgsMeshLayer::onDatasetGroupsAdded` which would just reassign all `assignDefaultStyleToDatasetGroup` to all dataset groups. But I would do this ONLY if we had to remove some extra uris on the way (because it would essentially remove all user settings).

vcloarec replied:
Hi,
ok, I will work on this signal/slot.

But just an idea : if the difficulty comes from extra URI which can have multiple dataset groups, what about a table who store all the dataset groups that are needed to be loaded (valid, no valid, compatible, no compatible,...) ? This table could store information which could be used to handle the dataset groups.
The entry key could be the index group for Qgis and the stored information could be the uri, the index for MDAL if any (if valid and compatible),... . Access method of the provider will used this table to retrieve the MDAL corresponding dataset group and return it, or return void dataset group if no valid or not compatible. This table could be refresh when reload is called.
This table could be the base for a table model associate to a table view in the mesh properties dialog.
Now, it is just an idea... but it could be not complicated to implement.

PeterPetrik replied:
whatever works :) but the MDAL group index is tricky, since if you load extraUris in different order it is different index. We would probably need that global id for groups some time anyways.

vcloarec replied:
ok, done.
I have added a way to avoid that non extras dataset groups have user setting removed.

I look on the MDAL code, and it seems the extra dataset groups are always added at the end of the DatasetGroups So, I don't understand why MDAL group index is tricky? The provider load extraUris always in the same order, so the the dataset groups have always the same order, no ?

PeterPetrik replied:
yes

PeterPetrik reviewed (changes requested):
your changes are good, I am just wondering now if the GUI is properly refreshed

vcloarec commented:
> I think you need to emit `QgsMeshDataProvider::dataChanged` signal here to notify GUI to reload widgets and refresh canvas.

There is a problem :
`QgsDataProvider::dataChanged` is connect with `refreshMapCanvas` (line 12349 in qgisapp.cpp)

This method calls `QgsMapCanvas::refreshAllLayers` (line 9401 same file)

`QgsMapCanvas::refreshAllLayer()` calls `QgsMapLayer::reload() `

and `QgsMeshLayer::reload()` calls `QgsMeshDataProvider::reloadData()` 

making a **circular reference**

Is it really necessary to emit `QgsMeshDataProvider::dataChanged` from here? 
May the caller of `QgsMeshLayer::reload()` or `QgsMeshDataProvider::reloadData()` can keep the responsability to inform the data has changed when necessary ?

PeterPetrik commented:
@nyalldawson Hi, just to double check. `QgsMapLayer::reload()` is only used with forced reload of all data from the data provider, so it can be very "slow", right?

PeterPetrik commented on src/core/mesh/qgsmeshdataprovider.h:
```diff
@@ -556,6 +556,11 @@ class CORE_EXPORT QgsMeshDataProvider: public QgsDataProvider, public QgsMeshDat
     void datasetGroupsAdded( int count );
+
+
+    //! Emitted when one or some extras uris can't be loaded. The mesh layer will receive this signal and will treat the rendere settings
```
Can you just remove the parameter, since I do not thing that the logic you created is valid for all cases and I find it a bit confusing to understand. I suggest

```
//! Emitted when one or more dataset groups have been removed
void datasetGroupsRemoved();
```

PeterPetrik commented on src/providers/mdal/qgsmdalprovider.cpp:
```diff
@@ -241,6 +234,49 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
+
+      if ( MDAL_LastStatus() != MDAL_Status::None )
+      {
+        atLeastOneExtraNotLoaded |= true;
```
why not `atLeastOneExtraNotLoaded=true`

vcloarec replied:
Hi,

I have purposed this logic to avoid removing all the users renderer settings, to reduce this dommageable consequences. I supposed the count of dataset groups present in the mesh file couldn't change, and, in fact, maybe it could change as well... 
So, in the new commit i have just purposed, if  extras uris can't be loaded **or** dataset groups count of the mesh file has changed, all the renderer settings is removed.
It is a bit confusing and I think there still are some cases that are not cover.
Maybe not satisfying...
I don't have time anymore today to think about, I will do later.

vcloarec committed:
Fix QgsMeshDatasetGroupTreeModel::groupIndexToModelIndex and revert to old commit

vcloarec commented:
Hi,
Ok, thinking we are on the wrong way, I restart the reasoning :
What happen if a dataset group is invalid or incompatible : the datasetGroup is not loaded and not indexes with MDAL.
What are the consequences : QHash in the render settings contain the wrong indexes of the dataset groups which a have been loaded after the invalid dataset groups. So this datagroups will have not an expected display in the canvas.
It is bad ? maybe not, this alert the user there is something wrong and he can modify his dataset groups file. After all, if the files is corrupt or contain a incompatible mesh after it was loaded the first time in QGIS, it is not the fault of QGis, but a inattentive action from the user.

So I try something : load an extra dataset in QGis with my PR (running the application), change the data set to set the mesh incompatible and reload all layer --> QGis Crash. Ok it is a fault of QGis with my PR...

But what happen with the release 3.6.2 : I load a mesh, a extra dataset group from a file, I make it active, save the project, quit, set the dataset groups file incompatible, restart Qgis and open the project -->
QGis Crash : it is not only my PR fault.

I fixes the bug in my new commit (from `QgsMeshDatasetGroupTreeModel::groupIndexToModelIndex`).
In addition to this fix, I don't remove the invalid or incompatible uris. So the user can retrieve his data after he has fixed his files. And not need anymore the signal/slot for removing dataset groups.

It seems the reload work fine with this modification, in any case, no less than the last release. Maybe is it enough ?

For a removing dataset groups mechanism, it is another story ...

PeterPetrik commented:
related to https://issues.qgis.org/issues/20124

PeterPetrik commented:
> I fixes the bug in my new commit (from `QgsMeshDatasetGroupTreeModel::groupIndexToModelIndex`).
> In addition to this fix, I don't remove the invalid or incompatible uris. So the user can retrieve his data after he has fixed his files. And not need anymore the signal/slot for removing dataset groups.

Thank you for looking into this. The reload function and removal of datasets is tricky fix, so I am glad that you really focused on getting this right. So far I am fine with the code in this PR, but I have found another crash.

You probably need a signal for QGIS app library to restart the state for non-existent layers. 
As I suggested create signal `void QgsMeshDataProvider::datasetGroupsRemoved()` and listen to it in GUI. As minimum you need to invalidate the renderer cache and active renderer dataset. 

to demonstrate the issue with the code:

1. Open QGIS, load 2dm file and one extra dataset (e.g. Depth)
2. render the extra dataset so you see its data on canvas
3. open 2dm file in text editor and add one vertex
4. click F5 (reload)
5. crash in `QgsMeshLayerRenderer::copyScalarDatasetValues` or vector equivalent 

I think it would be enough to clean renderer cache and set `mRendererSettings.activeScalarDataset()` and `acviteVectorDataset` to invalid dataset index.

vcloarec commented:
Hi,
I tried to reproduce the issue with my PR, but the crash doesn't appear.
The only consequences is that the extra dataset is not  display in the canvas, either in the mesh layer properties, as expected.

PeterPetrik commented:
> Hi,
> I tried to reproduce the issue with my PR, but the crash doesn't appear.
> The only consequences is that the extra dataset is not display in the canvas, either in the mesh layer properties, as expected.

hmm, that is strange since it crashed on my build with your branch... so when you have scalar and vector dataset loaded and you have switched on rendering of vector arrows there is no crash at all?

vcloarec commented:
no, I use the file in test/testdata/mesh/. And no crash, either with vector ...

PeterPetrik commented:
ok in that case one last thing. can you please rebase interactively your branch into one commit with sensible commit message and  reference that it is fix to https://issues.qgis.org/issues/20124 ? thanks (best to use git rebase -i master command).

vcloarec commented:
ok
To be clean I closed this PR and replace it with a PR  and reference is with fix #20124

PeterPetrik commented:
ok thanks. actually it is cleaner to rebase existing branch and force push. It would be better to do it that way for future PRs, so we do not loose existing comments and all. https://github.com/qgis/QGIS/pull/9922

vcloarec commented:
Sorry for that. I understand. As I have make some "experiences" with my local repository, I was afraid it is not very clean, and maybe a reason that the PR didn't work with you.
